sam_view.c

```diff
@@ -560,211 +560,425 @@ static const char *copied_tags[] = { "RG", "BC", "QT", NULL };
+        print_error_errno("Cannot read file \"%s\"", opts->fn_input);
+        free(state);
+        return true;
+    }
+    if (hts_set_opt(state->fp, CRAM_OPT_REQUIRED_FIELDS,
                     SAM_QNAME | SAM_FLAG | SAM_SEQ | SAM_QUAL)) {
         fprintf(stderr, "Failed to set CRAM_OPT_REQUIRED_FIELDS value\n");
-        return 1;
+        free(state);
+        return true;
     }
-    if (hts_set_opt(fp, CRAM_OPT_DECODE_MD, 0)) {
+    if (hts_set_opt(state->fp, CRAM_OPT_DECODE_MD, 0)) {
         fprintf(stderr, "Failed to set CRAM_OPT_DECODE_MD value\n");
-        return 1;
+        free(state);
+        return true;
+    }
+    if (opts->fnse) {
+        state->fpse = fopen(opts->fnse,"w");
+        if (state->fpse == NULL) {
+            print_error_errno("Cannot write to singleton file \"%s\"", opts->fnse);
+            free(state);
+            return true;
+        }
     }
 
-    h = sam_hdr_read(fp);
-    if (h == NULL) {
-        fprintf(stderr, "Failed to read header for \"%s\"\n", argv[optind]);
-        return 1;
+    state->fpr[0] = stdout;
+    state->fpr[1] = stdout;
+    state->fpr[2] = stdout;
+
+    if (opts->fnr0) {
+        state->fpr[0] = fopen(opts->fnr0, "w");
+        if (state->fpr[0] == NULL) {
+            print_error_errno("Cannot write to r0 file \"%s\"", opts->fnr1);
+            free(state);
+            return true;
+        }
     }
-    b = bam_init1();
-    if (b == NULL) {
-        perror(NULL);
-        bam_hdr_destroy(h);
-        return 1;
+
+    if (opts->fnr1) {
+        state->fpr[1] = fopen(opts->fnr1, "w");
+        if (state->fpr[1] == NULL) {
+            print_error_errno("Cannot write to r1 file \"%s\"", opts->fnr1);
+            free(state);
+            return true;
+        }
     }
-    buf = NULL;
-    max_buf = 0;
 
-    fpse = NULL;
-    if (fnse) {
-        fpse = fopen(fnse,"w");
-        if (fpse == NULL) {
-            print_error_errno("Cannot write to singleton file \"%s\"", fnse);
-            bam_hdr_destroy(h);
-            return 1;
+    if (opts->fnr2) {
+        state->fpr[2] = fopen(opts->fnr2, "w");
+        if (state->fpr[2] == NULL) {
+            print_error_errno("Cannot write to r2 file \"%s\"", opts->fnr2);
+            free(state);
+            return true;
         }
     }
 
-    int64_t n_singletons = 0, n_reads = 0;
-    char* previous = NULL;
-    kstring_t linebuf = { 0, 0, NULL };
+    state->h = sam_hdr_read(state->fp);
+    if (state->h == NULL) {
+        fprintf(stderr, "Failed to read header for \"%s\"\n", opts->fn_input);
+        free(state);
+        return true;
+    }
+
+    *state_out = state;
+    return false;
+}
+
+static bool destroy_state(const bam2fq_opts_t *opts, bam2fq_state_t *state, int* status)
+{
+    bam_hdr_destroy(state->h);
+    check_sam_close(state->fp, opts->fn_input, "file", status);
+    if (state->fpse && !fclose(state->fpse)) { print_error_errno("Error closing singleton file \"%s\"", opts->fnse); return true; }
```

Comment:
This should be fclose, not !fclose.  I get errors about successfully closing the singleton file.